### PR TITLE
[filebeat][gcs] Fix for concurrency issues and context timeouts in the GCS input 

### PR DESCRIPTION
**Pull request by ShourieG**
## Type of change
- Bug

## What does this PR do?
1. It fixes a map concurrency issue which was revealed when testing the input at scale.
2. Fixes the issue were bucket_timeout was being applied to the bucket poll interval instead of individual bucket object read ops.
3. Fixes flakey test issue.

## Why is it important?
This relates to an SDH from one of our customers and fixes those issues. 
This also addresses any issues the input could have at scale.

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc`.

### CODE REVIEW

File: x-pack/filebeat/input/gcs/job.go [hunk 1]
```diff
@@ -109,12 +109,14 @@ func (j *job) do(ctx context.Context, id string) {
 			Fields:    fields,
 		}
 		event.SetID(objectID(j.hash, 0))
-		j.state.save(j.object.Name, j.object.Updated)
-		// locks while data is being published to avoid concurrent map read/writes
+		// locks while data is being saved and published to avoid concurrent map read/writes
 		j.mu.Lock()
-		if err := j.publisher.Publish(event, j.state.checkpoint()); err != nil {
+		j.state.save(j.object.Name, j.object.Updated)
+		cp, done := j.state.checkpointTxn()
```
Comment (efd6) [inline x-pack/filebeat/input/gcs/job.go:115]: I had more of a think about this. If we have 
```
// save, saves/updates the current state for cursor checkpoint
func (s *state) save(name string, lastModifiedOn time.Time) {
	_, done := s.saveForTx(name, lastModifiedOn)
	done()
}

// saveForTx updates and returns the current state checkpoint, locks the state
// and returns an unlock function, done. The caller must call done when
// cp is no longer needed in a locked state. done may not be called
// more than once.
func (s *state) saveForTx(name string, lastModifiedOn time.Time) (cp *Checkpoint, done func()) {
	s.mu.Lock()
	delete(s.cp.LastProcessedOffset, name)
	delete(s.cp.IsRootArray, name)
	if _, ok := s.cp.FailedJobs[name]; !ok {
		if len(s.cp.ObjectName) == 0 {
			s.cp.ObjectName = name
		} else if strings.ToLower(name) > strings.ToLower(s.cp.ObjectName) {
			s.cp.ObjectName = name
		}

		if s.cp.LatestEntryTime.IsZero() {
			s.cp.LatestEntryTime = lastModifiedOn
		} else if lastModifiedOn.After(s.cp.LatestEntryTime) {
			s.cp.LatestEntryTime = lastModifiedOn
		}
	} else {
		// clear entry if this is a failed job
		delete(s.cp.FailedJobs, name)
	}
	return s.cp, func() { s.mu.Unlock() }
}
```
then j.state.save/j.state.checkpointTxn merge into a single call/lock with j.state.saveForTxn and j.mu should no longer be needed.

File: CHANGELOG.next.asciidoc [hunk 2]
```diff
@@ -124,6 +124,7 @@ https://github.com/elastic/beats/compare/v8.7.1\...main[Check the HEAD diff]
 - Fix crash when loading azurewebstorage cursor with no partially processed data. {pull}35433[35433]
 - RFC5424 syslog timestamps with offset 'Z' will be treated as UTC rather than using the default timezone. {pull}35360[35360]
 - [system] sync system/auth dataset with system integration 1.29.0. {pull}35581[35581]
+- [Gcs Input] - Enhancement and bugfix for concurrency issues, flakey tests and context timing out. {pull}35605[35605]
```
Comment (efd6) [inline CHANGELOG.next.asciidoc:127]:
```suggestion
- [GCS Input] - Enhancement and bugfix for concurrency issues, flakey tests and context timing out. {pull}35605[35605]
```
but I think these should be broken out into separate lines that clearly describe each change (probably the enhancement should be a separate PR).

File: x-pack/filebeat/input/gcs/input_stateless.go
Comment (efd6) [inline x-pack/filebeat/input/gcs/input_stateless.go:1]: You may as well keep the linter happy here too.

File: x-pack/filebeat/input/gcs/job.go (hunk 1 above)
Comment (efd6) [reply x-pack/filebeat/input/gcs/job.go:115]: Similar change for `savePartial` for below. It looks like `save` (and `savePartial`) is always used before a publish, so really we don't need the self-re-unlocking variant, though I'd use the new name for clarity.

**Comment by ShourieG** [comment]
@efd6 , I've addressed the changes and updated the PR

File: x-pack/filebeat/input/gcs/job.go [hunk 3]
```diff
@@ -216,19 +213,21 @@ func (j *job) readJsonAndPublish(ctx context.Context, r io.Reader, id string) er
 		// updates the offset after reading the file
 		// this avoids duplicates for the last read when resuming operation
 		offset = dec.InputOffset()
+		// locks while data is being saved and published to avoid concurrent map read/writes
+		var done func()
+		var cp *Checkpoint
```
Comment (efd6) [inline x-pack/filebeat/input/gcs/job.go:218]: nit
```suggestion
	var (
		cp   *Checkpoint
		done func()
	)
```

File: x-pack/filebeat/input/gcs/scheduler.go [hunk 4]
```diff
@@ -92,40 +85,54 @@ func (l *limiter) release() {
...
+				job.do(ctx, id)
 			}()
 		}
 
+		s.log.Debugf("scheduler: total objects read till now: %d", numObs)
+		s.log.Debugf("scheduler: total jobs scheduled till now: %d", numJobs)
+		if len(jobs) > 0 {
+			s.log.Debugf("scheduler: first job in current batch: %s", jobs[0].Name())
+			s.log.Debugf("scheduler: last job in current batch: %s", jobs[len(jobs)-1].Name())
+		}
```
Comment (efd6) [inline x-pack/filebeat/input/gcs/scheduler.go:129]: These could be bundled into two `s.log.Debugf` calls so that the log level isn't checked 2×.

File: x-pack/filebeat/input/gcs/scheduler.go [hunk 5]
```diff
@@ -216,6 +223,8 @@ func (s *scheduler) addFailedJobs(ctx context.Context, jobs []*job) []*job {
 		jobMap[j.Name()] = true
 	}
 
+	s.log.Debugf("scheduler: %d failed jobs found", len(s.state.checkpoint().FailedJobs))
```
Comment (efd6) [inline x-pack/filebeat/input/gcs/scheduler.go:226]: This is racey (actually the `state.checkpoint` method is a race footgun).

File: CHANGELOG.next.asciidoc (hunk 2 above)
Comment (efd6) [reply CHANGELOG.next.asciidoc]: This doesn't explain what is the enhancement and what is the bugfix. Ideally when there is a subtle bugfix, there should be no concomitant enhancememt.

File: x-pack/filebeat/input/gcs/scheduler.go (hunk 5 above)
Comment (ShourieG) [reply x-pack/filebeat/input/gcs/scheduler.go:226]: @efd6, should we lock this also then ?

File: CHANGELOG.next.asciidoc (hunk 2 above)
Comment (ShourieG) [reply CHANGELOG.next.asciidoc]: Sure, I'll break it into different statements. As for the PR I agree they should be separate usually but since this was more of a minor semantic change of passing the context instead of storing in structs and re enabling the tests in addition to the bug fixes, I thought I'd keep them in so it can be reviewed in one go.

File: x-pack/filebeat/input/gcs/scheduler.go (hunk 5 above)
Comment (ShourieG) [reply x-pack/filebeat/input/gcs/scheduler.go:226]: and I checked that in every place where s.state.checkpoint() is called are before any jobs are spawned

**Comment by ShourieG** [comment]
@efd6, Updated the changes and left comment regarding https://github.com/elastic/beats/pull/35605#discussion_r1209776641

File: x-pack/filebeat/input/gcs/scheduler.go (hunk 5 above)
Comment (efd6) [reply x-pack/filebeat/input/gcs/scheduler.go]: It's worth understanding whether this can be concurrently updated here.

**Review by efd6 — APPROVED** [review]
I think I'm OK with the concurrency fix, mod the unlocked use of the `*Checkpoint` returned by `*state.checkpoint()`. I think that method needs documentation clearly stating that it's not safe for concurrent use and that the returned value can not be used while the receiver in operating concurrently. Similar for `*state.setCheckpoint()`.

Overall, I'd like to see this broken into separate PRs.

Please wait for @andrewkroh

File: CHANGELOG.next.asciidoc [hunk 6]
```diff
@@ -124,6 +124,11 @@ https://github.com/elastic/beats/compare/v8.7.1\...main[Check the HEAD diff]
 - Fix crash when loading azurewebstorage cursor with no partially processed data. {pull}35433[35433]
 - RFC5424 syslog timestamps with offset 'Z' will be treated as UTC rather than using the default timezone. {pull}35360[35360]
 - [system] sync system/auth dataset with system integration 1.29.0. {pull}35581[35581]
+- [GCS Input] - Fixed an issue where bucket_timeout was being applied to the entire bucket poll interval and not individual 
+   bucket object read operations. Fixed a map write concurrency issue arising from data races when using a high number of workers. 
+   Fixed the flaky tests that were present in the GCS test suit. {pull}35605[35605]
+- [GCS Input] - Enhanced the input by adding debug logs and improved the context passing mechanism by removing them from struct params 
+   and passing them as function arguments. {pull}35605[35605]
```
[the remark this answers is not in the document]
Comment (ShourieG) [reply CHANGELOG.next.asciidoc:131]: I'd like the bug fixes to be in the same PR so that they can be fast tracked as they relate to an urgent SDH. I can separate out the enhancements into a separate PR(context and logging).

**Comment by ShourieG** [comment]
@efd6, I've removed the enhancements from this PR, I will put up another PR post merging of this PR which will contain those enhancements.

**Comment by efd6** [comment]
Thanks. Please still wait for Andrew.

File: CHANGELOG.next.asciidoc [hunk 7]
```diff
@@ -124,6 +124,9 @@ https://github.com/elastic/beats/compare/v8.7.1\...main[Check the HEAD diff]
 - Fix crash when loading azurewebstorage cursor with no partially processed data. {pull}35433[35433]
 - RFC5424 syslog timestamps with offset 'Z' will be treated as UTC rather than using the default timezone. {pull}35360[35360]
 - [system] sync system/auth dataset with system integration 1.29.0. {pull}35581[35581]
+- [GCS Input] - Fixed an issue where bucket_timeout was being applied to the entire bucket poll interval and not individual 
+   bucket object read operations. Fixed a map write concurrency issue arising from data races when using a high number of workers.
```
Comment (andrewkroh) [inline CHANGELOG.next.asciidoc:128]: IIRC the changelog script expects messages to be on a single line.

**Comment by ShourieG** [comment]
@andrewkroh does the PR look good for merging at the moment ?

**Review by andrewkroh — COMMENTED** [review]
This looks like a good improvement in address the concurrency issues. So if you wanted to go ahead as it that's good with me.

But I think there may be some correctness issues w.r.t. the comment I raised. This could be apparent during shutdown where events in the queue are discarded, but the state was already updated with the assumption that the events where written to ES.

File: x-pack/filebeat/input/gcs/job.go [hunk 8]
```diff
@@ -216,19 +213,23 @@ func (j *job) readJsonAndPublish(ctx context.Context, r io.Reader, id string) er
...
+			cp, done = j.state.savePartialForTx(j.object.Name, offset+relativeOffset)
 		}
-		// locks while data is being published to avoid concurrent map read/writes
-		j.mu.Lock()
-		if err := j.publisher.Publish(evt, j.state.checkpoint()); err != nil {
+		if err := j.publisher.Publish(evt, cp); err != nil {
 			j.log.Errorw("job encountered an error", "gcs.jobId", id, "error", err)
 		}
-		j.mu.Unlock()
+		// unlocks after data is saved and published
```
[the remark this answers is not in the document]
Comment (ShourieG) [reply x-pack/filebeat/input/gcs/job.go:231]: @andrewkroh thanks for the feedback, I will revisit this when I do some planned feature enhancements for the gcs input in the near future, will have to separate out the cursor publish from the event publish to properly implement this feature. For now I'm merging then PR since it's related to an sdh.
